src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -740,27 +736,52 @@ private static void RegisterGrainReferenceSerializers(Type type)
                 GrainReference.SerializeGrainReference,
                 (expected, stream) =>
                 {
-                    ConstructorInfo constructor;
-                    if (expected.IsConstructedGenericType)
+                    Func<GrainReference, GrainReference> ctorDelegate;
+                    var deserialized = (GrainReference)GrainReference.DeserializeGrainReference(expected, stream);
+                    if (expected.IsConstructedGenericType == false)
                     {
-                        var referenceType = type.MakeGenericType(expected.GenericTypeArguments);
-                        constructor =
-                            referenceType.GetConstructor(
-                                BindingFlags.CreateInstance | BindingFlags.NonPublic | BindingFlags.Instance,
-                                null,
-                                new[] { typeof(GrainReference) },
-                                null);
+                        return defaultCtorDelegate(deserialized);
                     }
-                    else
+
+                    if (!grainRefConstructorDictionary.TryGetValue(expected, out ctorDelegate))
                     {
-                        constructor = nonGenericConstructor;
+                        ctorDelegate = CreateGrainRefConstructorDelegate(type, expected.GenericTypeArguments);
+                        grainRefConstructorDictionary.TryAdd(expected, ctorDelegate);
                     }
 
-                    var deserialized = (IAddressable)GrainReference.DeserializeGrainReference(expected, stream);
-                    return (IAddressable)constructor.Invoke(new object[] { deserialized });
+                    return ctorDelegate(deserialized);
                 });
         }
 
+        private static Func<GrainReference, GrainReference> CreateGrainRefConstructorDelegate(Type type, Type[] genericArgs)
+        {
+            if (type.IsGenericType)
+            {
+                if (type.IsConstructedGenericType == false && genericArgs == null)
+                {
+                    return null;
+                }
+
+                type = type.MakeGenericType(genericArgs);
+            }
+
+            var constructor =
+                type.GetConstructor(
+                    BindingFlags.CreateInstance | BindingFlags.NonPublic | BindingFlags.Instance,
```

Comment:
will this work even if the constructor is actually non-public, as the bit flags specifies? I know it works for reflection, but I don't know how it behaves with expression trees. And do we actually have grain references with non-public constructors?